src/librustc_typeck/collect.rs

```diff
@@ -1085,12 +1085,16 @@ fn type_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
 
         NodeField(field) => icx.to_ty(&field.ty),
 
-        NodeExpr(&hir::Expr { node: hir::ExprClosure(.., gen), .. }) => {
+        NodeExpr(&hir::Expr { node: hir::ExprClosure(.., gen), span, .. }) => {
             if gen.is_some() {
                 let hir_id = tcx.hir.node_to_hir_id(node_id);
                 return tcx.typeck_tables_of(def_id).node_id_to_type(hir_id);
             }
 
+            if !tcx.has_typeck_tables(def_id) {
+                span_err!(tcx.sess, span, E0912, "closures cannot be constants");
```

Comment:
Nit: closures *can* be constants:

```rust
const FOO: fn() = || println!("Gah!");
```
Maybe there's some other error message that could be more specific?

Reply:
True, what about `expected numerical constant, found closure`?

Reply:
That sounds much better to me, thanks!

Reply:
Do we need this check at all? I'd assume if we left it out, we'd get the real typeck error.